### PR TITLE
feat(recall): code-graph steering, richer indexing, and a restart-free enable/disable toggle

### PR DESCRIPTION
**Pull request by ziadeh**
## Summary

Hardens Recall's code-graph pillar and closes two lifecycle gaps.

### Code graph
- **Steer the agent to actually use the graph** — session provisioning now teaches agents to reach for `graph_search` / `graph_node` / `get_neighbors` / `impact_of` / `shortest_path` instead of grepping; graph token selection drops stopwords so the specific noun wins the ranking.
- **Verbatim symbol source** — `graph_node` (and `graph_search?source=1`) return a symbol's definition source inline, so the agent can read a definition without opening the file.
- **Broader language coverage** — the indexer now handles ESM/CJS variants (`.mts` / `.cts` / `.mjs` / `.cjs`) and Python alongside TS/TSX/JS/JSX.
- **Staleness flags** — graph results flag files changed since the last index, so line ranges shifted by live edits are visible to the consumer.

### Lifecycle fixes
- **Guaranteed memory injection at session start** — the Session Brief lands even when the spawn-time fetch loses the race.
- **Restart-free Recall toggle** — flipping the Recall master switch previously required relaunching the app:
  - Session spawn now reads the live `recallEnabled` flag (fail-soft) and, when off, actively removes the managed `.mcp.json` entry and skill folder. Skill removal is ownership-checked so a user-authored skill at the same path survives.
  - The memory API and graph endpoints refuse with 403 while Recall is off, so already-running sessions (whose MCP config can't be hot-swapped) and stale `.mcp.json` entries picked up outside Mission Control go dead immediately.
  - The task brief endpoint intentionally stays open and serves an empty 200 when disabled — the spawn path relies on it to strip stale managed blocks from agents' auto-load files.

## Testing
- Full vitest suite: 175 files / 1580 tests green (33 new tests: MCP-entry removal, skill removal + ownership sparing, settings fetch fail-soft paths, and 403 behavior across every gated endpoint).
- `tsc --noEmit` for both the app and `electron/tsconfig.json`, plus eslint, all clean.

### CODE REVIEW

File: src/server/services/code-graph-source.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+ * escapes the project root (a stale or corrupt row must never become an
+ * arbitrary-file read).
+ */
+export function readNodeSource(
+  root: string,
+  node: Pick<GraphNodeView, "kind" | "filePath" | "startLine" | "endLine">,
+  maxLines: number,
+): GraphNodeSource | null {
+  const abs = path.resolve(root, node.filePath);
+  if (abs !== root && !abs.startsWith(root + path.sep)) return null;
```
Comment (webdevcody) [inline src/server/services/code-graph-source.ts:35]: **Low severity — symlink escape not covered by the root-prefix check.**

`readNodeSource` guards against path traversal by comparing the *lexical* resolved path against `root` (`path.resolve` + `startsWith`), but it never resolves symlinks. If a symlink lives inside the project root and points outside it (e.g. committed accidentally, or created by a prior malicious commit/build step), `abs.startsWith(root + sep)` still passes and `fs.readFileSync(abs, ...)` will happily follow the symlink and return content from outside the project root through the `graph_node` / `graph_search?source=1` API.

This is a narrow risk (requires a symlink already present under the indexed root, and the indexer would have had to pick it up as a "file" node in the first place), but since the whole point of this function's guard is "a stale or corrupt row must never become an arbitrary-file read" per the docstring, it's worth closing the gap: compare `fs.realpathSync(abs)` against `fs.realpathSync(root)` (or at least stat with `lstat`/`throwIfNoEntry` and reject non-regular-file targets reached via a symlink) rather than trusting the raw resolved path.

File: src/shared/agent-hooks.ts [hunk 2]
```diff
@@ -191,10 +192,34 @@ function buildManagedGroup(
...
 
+// A hook command that posts to Mission Control's own hook endpoint is ours by
+// construction — $MC_TASK_ID / $MC_API_URL only exist inside MC-spawned
+// sessions. Early installer versions wrote these entries WITHOUT the marker, so
+// marker-only filtering let them accumulate as duplicates: each fires an extra
+// POST per event, and for injectContext events the stale (stdout-discarding)
+// copies race the managed hook server-side and swallow its one-shot output.
+function isMissionControlCommand(command: unknown): boolean {
+  return (
+    typeof command === "string" &&
```
Comment (webdevcody) [inline src/shared/agent-hooks.ts:203]: **Informational — legacy-hook sweep heuristic is a substring match, not an identity check.**

`isMissionControlCommand` treats *any* hook command string containing both `/api/hooks/` and `MC_TASK_ID` as "ours" and therefore eligible for removal/replacement by `installAgentHooks`/`removeManagedEvents`. Since these are just `String.includes` checks on user-editable JSON (Claude Code / Cursor hook config), a user-authored hook that happens to reference those two substrings (e.g. a wrapper script comment, a curl command hitting a *different* `/api/hooks/` path with an env var literally named `MC_TASK_ID`) would be silently treated as a legacy Mission Control entry and dropped on the next spawn.

This is a low-likelihood collision and the intent (sweeping up pre-marker duplicates) is reasonable, but given the marker-based path (`group[MARKER]`) already exists and is precise, it might be worth tightening the legacy fallback — e.g. also requiring the command to reference the specific hook endpoint host/port from `mcEnv`, or the exact script name Mission Control writes — so a coincidental match in a hand-written hook can't be swept up.

**Review by webdevcody — COMMENTED** [review]
## Security review summary

I went through the full diff (40 files, ~2.2k additions) focused on malicious/insecure code — path handling, auth/access-control changes, code execution, and injection surfaces.

**What I checked closely:**
- `code-graph-source.ts` (new: reads arbitrary node source off disk at query time) — has an explicit root-escape guard + tests, but see inline comment on a symlink edge case.
- `code-graph.controller.ts` / `project-memory.controller.ts` — new 403 gating (`requireRecallOn` / `requireGraphOn`) is applied consistently across every graph and memory endpoint; `recall-settings.ts` defaults `enabled: false` (fail-closed) so the master switch fails safe.
- `ensure-recall-mcp.ts` / `ensure-recall-skill.ts` — the new remove-on-disable paths only ever touch two fixed, well-known keys (`recall` / `recall-graph`) in `.mcp.json`, and skill removal is gated on an ownership check (`SKILL.md` must self-identify as Mission Control's) before `rmSync`, so a user-authored skill at the same path survives.
- `recall-enabled.ts` — calls the local MC API with the session's own bearer token, times out at 1.5s, fails soft (null) rather than throwing.
- `code-graph.repo.ts` `searchNodesFuzzy` — new SQL uses Drizzle's parameterized `sql` template (`instr(lower(name), ${q})`), not string concatenation, so no SQL injection.
- `code-graph-extract.ts` Python AST walk, `code-graph-indexer.ts` import resolution, `code-graph-wasm.ts` grammar loading — plain parsing/traversal, no `eval`/`exec`/`child_process` anywhere in the diff.
- No secrets are written to disk: the `.mcp.json` entry uses Claude Code's own `${MC_API_URL:-}`-style env expansion rather than embedding the token literally.

**Findings posted inline (both low severity/informational, not blocking):**
1. `code-graph-source.ts:35` — the project-root escape check is a lexical `path.resolve` + `startsWith` comparison and doesn't resolve symlinks, so a symlink planted inside the project root pointing outside it could be followed by `readFileSync` and leak content outside the root through `graph_node`/`graph_search?source=1`. Narrow (requires a pre-existing symlink under the indexed root) but worth a `realpathSync` comparison to fully match the stated intent ("must never become an arbitrary-file read").
2. `src/shared/agent-hooks.ts:203` — the legacy-hook sweep (`isMissionControlCommand`) identifies "ours" by substring-matching `/api/hooks/` + `MC_TASK_ID` in a hook command, rather than a precise marker. A coincidental match in a user-authored hook would get silently swept on the next spawn. Low likelihood, but worth tightening since the precise marker-based path already exists.

No signs of intentionally malicious code, backdoors, credential exfiltration, or unsafe command execution. The gating/ownership checks for the new toggle-off behavior are well thought through and covered by tests.

## Confidence score: 4/5

Solid, well-tested change with consistent access-control gating and no exec/eval/injection issues found. Not a 5 because of the symlink-based path-escape gap in the new source-reading endpoint (`code-graph-source.ts`) — I'd want that tightened (or explicitly accepted as out-of-scope, given it requires an existing malicious symlink in the repo) before treating file-content-returning endpoints as fully hardened.

(Per review process: not merging this PR myself regardless of score.)
